### PR TITLE
issue#40

### PR DESCRIPTION
**Pull request by jubaldo**

### CODE REVIEW

File: src/ctree-element-screen/ctree-segment-container.html [hunk 1]
```diff
@@ -187,13 +187,15 @@
 
       previous() {
         if (this._variationIndex <= 0) return;
-
+        var component = this.shadowRoot.querySelector('#component');
+        component.editing= false;
```
Comment (CSchmitz81) [inline src/ctree-element-screen/ctree-segment-container.html]: It's a minor point, but I like to stress with developers that they should use consistent formatting.  This is missing a space before the `=`.

File: src/ctree-segments/ctree-segment-behavior.html [hunk 2]
```diff
@@ -30,6 +30,7 @@
 <link rel="import" href="../../bower_components/polymer/polymer.html">
 <link rel="import" href="../../bower_components/polymer/lib/utils/async.html">
 
+
```
Comment (CSchmitz81) [inline src/ctree-segments/ctree-segment-behavior.html]: Another consistency point that we don't usually have 2 empty lines after the imports.

File: src/ctree-segments/ctree-segment-behavior.html [hunk 3]
```diff
@@ -123,6 +128,26 @@
     Polymer.Async.microTask.cancel(this.__editUpdateHandle);
   }
 
+  _properExit(editing){
```
Comment (CSchmitz81) [inline src/ctree-segments/ctree-segment-behavior.html]: More consistency, missing a space before `{`

File: src/ctree-segments/ctree-segment-behavior.html [hunk 4]
```diff
@@ -123,6 +128,26 @@
...
+    if(!editing && this.validChangeMade){
```
Comment (CSchmitz81) [inline src/ctree-segments/ctree-segment-behavior.html]: More consistency, spaces before `(` and `{`.

File: src/ctree-segments/ctree-segment-image.html [hunk 5]
```diff
@@ -163,10 +166,11 @@
 
       _editSaveClicked() {
         super._addSegmentVariation(this.$.image.src);
+        super._closure();
```
Comment (CSchmitz81) [inline src/ctree-segments/ctree-segment-image.html]: I think you can use `this._closure()` instead of `super._closure()`, can't you?  Usually `super` is only used when you're overriding a function, because you don't want to short circuit your ability to override the functionality.

File: src/ctree-element-screen/ctree-segment-container.html [hunk 6]
```diff
@@ -187,13 +187,15 @@
 
       previous() {
         if (this._variationIndex <= 0) return;
-
+        var component = this.shadowRoot.querySelector('#component');
```
Comment (CSchmitz81) [inline src/ctree-element-screen/ctree-segment-container.html]: Hmm, I know this is done all over in this class (and I'm the one who did it), but it's really a lot of copy-paste code, which always is a good sign of an opportunity for improvement.  It's up to you, what you do (if anything), but a simple improvement would be to make a simple function to get the component.  A more sophisticated improvement would be to do something consistent with Polymer and see if we can set `this.$.component` when we create the component ID in `_segmentChanged`, which would prevent us from needing to use the more complicated (and inconsistent) method of calling querySelector.  That said, you'd need to test that carefully, because I'm not sure if there would be any negative implication of doing that (or maybe we don't even need to set it, maybe `this.$` automatically would work once the ID is set).

File: src/ctree-element-screen/ctree-segment-container.html [hunk 7]
```diff
@@ -187,18 +187,24 @@
...
 
       next() {
         if (!this.segment || !this.segment.variations || this._variationIndex >= this.segment.variations.length - 1) return;
-        var component = this.shadowRoot.querySelector('#component');
-        component.editing= false;
-        this._variationIndex++;
+        this._componentOff(1);
       }
 
+      _componentOff(incDec){
```
Comment (CSchmitz81) [inline src/ctree-element-screen/ctree-segment-container.html]: I know we talked about this in person, but just to document it here I should have been more clear about what I was referring to.  I was specifically talking about putting `this.shadowRoot.querySelector('#component');` in a helper function so we don't have to keep writing it.  That's only if we can't inject the dynamically created object into `this.$.component`, which would be the best case scenario.

Comment (CSchmitz81) [reply src/ctree-element-screen/ctree-segment-container.html]: Also, pay attention to code format here ;-)

File: src/ctree-element-screen/ctree-segment-container.html [hunk 8]
```diff
@@ -212,6 +216,7 @@
             parent.removeChild(child);
           }
           if (type) {
+            this.$.component = this.getComponent();
```
Comment (CSchmitz81) [inline src/ctree-element-screen/ctree-segment-container.html]: It looks like this is in the wrong place, because you're getting the component before creating it.  Also, shouldn't we only set the component in `this.$.component` after the component is created and the ID is set.  That way we don't need the `getComponent` (we can just inline that code here, where we create it).

File: src/ctree-element-screen/ctree-segment-container.html [hunk 9]
```diff
@@ -294,6 +298,7 @@
           component.set('data', variation.data);
           // TODO: don't set height percentage, let full image size be visible
           component.set('heightPercent', '56.25%');
+          this.$.component = this.getComponent();
```
Comment (CSchmitz81) [inline src/ctree-element-screen/ctree-segment-container.html]: I'm not sure why `this.$.component` is set here.  Shouldn't it only be set when the component is created.

File: src/ctree-element-screen/ctree-segment-container.html [hunk 10]
```diff
@@ -187,16 +187,20 @@
...
 
       next() {
         if (!this.segment || !this.segment.variations || this._variationIndex >= this.segment.variations.length - 1) return;
-
+        this.$.component.editing = false;
         this._variationIndex++;
       }
 
+      getComponent() {
+        return this.shadowRoot.querySelector("#component");
```
Comment (CSchmitz81) [inline src/ctree-element-screen/ctree-segment-container.html]: If you're updating references to ` this.shadowRoot.querySelector("#component")` I suggest doing a find and replace.  It's always a good practice to change similar things when possible (less time investment for improving the code).

**Comment by CSchmitz81** [comment]
@jubaldo I added a few comments where I wasn't sure why you did somethings, and a tip for when you're improving code.

**Comment by CSchmitz81** [comment]
@jubaldo I like the standardization, but it still isn't quite what I had in mind.  I think describing it isn't working, so rather than having you trying to continue guessing what I'm picturing, I put in a branch showing what I had in mind.  I did some minor testing to see if it worked at all, and it seems to, but I'll leave it to you to fully verify and merge if you agree with the changes.  Here's the diff: https://github.com/F4IF/ctree-demo/compare/develop_jubaldo...cschmitz/develop_jubaldo
